tests/all/component_model/call_hook.rs

```diff
@@ -0,0 +1,638 @@
+                    (realloc (func $i "realloc"))
+                )
+            )
+        )"#
+    );
+
+    let engine = Engine::default();
+    let component = Component::new(&engine, wat)?;
+    let linker = Linker::<State>::new(&engine);
+    let mut store = Store::new(&engine, State::default());
+    store.call_hook(sync_call_hook);
+    let inst = linker
+        .instantiate(&mut store, &component)
+        .expect("instantiate");
+
+    let export = inst
+        .get_typed_func::<(&[u8],), (WasmStr,)>(&mut store, "list8-to-str")
+        .expect("looking up `list8-to-str`");
+
+    let message = String::from("hello, world!");
+    let res = export.call(&mut store, (message.as_bytes(),))?.0;
+    let result = res.to_str(&store)?;
+    assert_eq!(&message, &result);
+
+    assert_eq!(store.data().calls_into_wasm, 2);
+    assert_eq!(store.data().returns_from_wasm, 2);
+
+    export.post_return(&mut store)?;
+
+    // There is one host call for the host-side realloc, and then two wasm calls for both the
+    // `list8-to-str` call and the guest realloc call for the list argument.
+    let s = store.into_data();
+    assert_eq!(s.calls_into_host, 1);
+    assert_eq!(s.returns_from_host, 1);
+    assert_eq!(s.calls_into_wasm, 2);
+    assert_eq!(s.returns_from_wasm, 2);
+
+    Ok(())
+}
+
+// Call a guest function that also defines a post-return.
+#[test]
+fn call_func_with_post_return() -> Result<()> {
+    let wat =
+        r#"(component
+            (core module $m
+                (func (export "roundtrip"))
+                (func (export "post-return"))
+            )
+            (core instance $i (instantiate $m))
+
+            (func (export "export")
+                (canon lift
+                    (core func $i "roundtrip")
+                    (post-return (func $i "post-return"))
+                )
+            )
+        )"#;
+
+    let engine = Engine::default();
+    let component = Component::new(&engine, wat)?;
+    let linker = Linker::<State>::new(&engine);
+    let mut store = Store::new(&engine, State::default());
+    store.call_hook(sync_call_hook);
+    let inst = linker
+        .instantiate(&mut store, &component)
+        .expect("instantiate");
+
+    let export = inst
+        .get_typed_func::<(), ()>(&mut store, "export")
+        .expect("looking up `export`");
+
+    export.call(&mut store, ())?;
+
+    // Before post-return, there will only have been one call into wasm.
+    assert_eq!(store.data().calls_into_wasm, 1);
+    assert_eq!(store.data().returns_from_wasm, 1);
+
+    export.post_return(&mut store)?;
+
+    // There are no host calls in this example, but the post-return does increment the count of
+    // wasm calls by 1, putting the total number of wasm calls at 2.
+    let s = store.into_data();
+    assert_eq!(s.calls_into_host, 0);
+    assert_eq!(s.returns_from_host, 0);
+    assert_eq!(s.calls_into_wasm, 2);
+    assert_eq!(s.returns_from_wasm, 2);
```

Comment:
Here are the assertions about `post-return`, you  can see that the wasm calls increment after the `.post_return()` call.